substrate/client/rpc-spec-v2/src/transaction/transaction_broadcast.rs

```diff
@@ -0,0 +1,292 @@
+	/// Executor to spawn subscriptions.
+	executor: SubscriptionTaskExecutor,
+	/// The brodcast operation IDs.
+	broadcast_ids: Arc<RwLock<HashMap<String, BroadcastState>>>,
+}
+
+/// The state of a broadcast operation.
+struct BroadcastState {
+	/// Handle to abort the running future that broadcasts the transaction.
+	handle: AbortHandle,
+}
+
+impl<Pool, Client> TransactionBroadcast<Pool, Client> {
+	/// Creates a new [`TransactionBroadcast`].
+	pub fn new(client: Arc<Client>, pool: Arc<Pool>, executor: SubscriptionTaskExecutor) -> Self {
+		TransactionBroadcast { client, pool, executor, broadcast_ids: Default::default() }
+	}
+
+	/// Generate an unique operation ID for the `transaction_broadcast` RPC method.
+	pub fn generate_unique_id(&self) -> String {
+		let generate_operation_id = || {
+			// The lenght of the operation ID.
+			const OPERATION_ID_LEN: usize = 16;
+
+			rand::thread_rng()
+				.sample_iter(Alphanumeric)
+				.take(OPERATION_ID_LEN)
+				.map(char::from)
+				.collect::<String>()
+		};
+
+		let mut id = generate_operation_id();
+
+		let broadcast_ids = self.broadcast_ids.read();
+
+		while broadcast_ids.contains_key(&id) {
+			id = generate_operation_id();
+		}
+
+		id
+	}
+}
+
+/// Currently we treat all RPC transactions as externals.
+///
+/// Possibly in the future we could allow opt-in for special treatment
+/// of such transactions, so that the block authors can inject
+/// some unique transactions via RPC and have them included in the pool.
+const TX_SOURCE: TransactionSource = TransactionSource::External;
+
+#[async_trait]
+impl<Pool, Client> TransactionBroadcastApiServer<BlockHash<Pool>>
+	for TransactionBroadcast<Pool, Client>
+where
+	Pool: TransactionPool + Sync + Send + 'static,
+	Pool::Error: IntoPoolError,
+	Pool::Hash: Unpin,
+	<Pool::Block as BlockT>::Hash: Unpin,
+	Client: HeaderBackend<Pool::Block> + BlockchainEvents<Pool::Block> + Send + Sync + 'static,
+{
+	fn broadcast(&self, bytes: Bytes) -> RpcResult<Option<String>> {
+		let pool = self.pool.clone();
+
+		// The unique ID of this operation.
+		let id = self.generate_unique_id();
+
+		let mut best_block_import_stream =
+			Box::pin(self.client.import_notification_stream().filter_map(
+				|notification| async move { notification.is_new_best.then_some(notification.hash) },
+			));
+
+		let broadcast_transaction_fut = async move {
+			// There is nothing we could do with an extrinsic of invalid format.
+			let Ok(decoded_extrinsic) = TransactionFor::<Pool>::decode(&mut &bytes[..]) else {
+				return;
+			};
+
+			// Flag to determine if the we should broadcast the transaction again.
+			let mut is_done = false;
+
+			while !is_done {
+				// Wait for the last block to become available.
+				let Some(best_block_hash) =
+					last_stream_element(&mut best_block_import_stream).await
+				else {
+					return;
+				};
+
+				let mut stream = match pool
+					.submit_and_watch(best_block_hash, TX_SOURCE, decoded_extrinsic.clone())
+					.await
```

Comment:
We are effectively trying to resubmit the transaction to the next best block here; if we cannot make progress at this time:
- `submit_and_watch` directly returned a recoverable error (ImmediatelyDroped etc)
- `submit_and_watch` produced an event that may be recoverable (Invalid / Dropped etc)

When we get the `Invalid` event, the transaction-pool will ban it for some time (default 30 mins).  In this case, the looping seems redundant unless chain developers manually pick another timeout.

However, I believe we can simplify this and try to submit the transaction only once. 
From the [transaction_unstable_broadcast](https://github.com/paritytech/json-rpc-interface-spec/blob/main/src/api/transaction_unstable_broadcast.md#transaction_unstable_broadcast), the phrasing states that the server tries to broadcast, but without any guarantees: 
> JSON-RPC server will try to propagate this transaction

This could be interpreted as we'll submit the transaction to the transaction pool now, and if it fails or is invalid we'll stop.

@skunert would love to get your thoughts on this 🙏

Reply:
Yeah sounds like its not enforced by the spec, but IMO the current implementation makes sense. If we can recover and retry in the next block, why not.